src/Workspaces/Core/Portable/Serialization/SerializerService_Asset.cs

```diff
@@ -21,29 +21,31 @@ namespace Microsoft.CodeAnalysis.Serialization
     /// </summary>
     internal partial class SerializerService
     {
-        public void SerializeSourceText(ITemporaryStorageWithName? storage, SourceText text, ObjectWriter writer, CancellationToken cancellationToken)
+        public void SerializeSourceText(SerializableSourceText text, ObjectWriter writer, CancellationToken cancellationToken)
         {
             cancellationToken.ThrowIfCancellationRequested();
-
-            writer.WriteInt32((int)text.ChecksumAlgorithm);
-            writer.WriteEncoding(text.Encoding);
-
-            // TODO: refactor this part in its own abstraction (Bits) that has multiple sub types
-            //       rather than using enums
-            if (storage != null && storage.Name != null)
+            if (text.Storage is not null)
             {
+                writer.WriteInt32((int)text.Storage.ChecksumAlgorithm);
+                writer.WriteEncoding(text.Storage.Encoding);
+
                 writer.WriteInt32((int)SerializationKinds.MemoryMapFile);
-                writer.WriteString(storage.Name);
-                writer.WriteInt64(storage.Offset);
-                writer.WriteInt64(storage.Size);
-                return;
+                writer.WriteString(text.Storage.Name);
+                writer.WriteInt64(text.Storage.Offset);
+                writer.WriteInt64(text.Storage.Size);
             }
+            else
+            {
+                RoslynDebug.AssertNotNull(text.Text);
 
-            writer.WriteInt32((int)SerializationKinds.Bits);
-            text.WriteTo(writer, cancellationToken);
+                writer.WriteInt32((int)text.Text.ChecksumAlgorithm);
+                writer.WriteEncoding(text.Text.Encoding);
+                writer.WriteInt32((int)SerializationKinds.Bits);
+                text.Text.WriteTo(writer, cancellationToken);
```

Comment:
on the fence if i would prefer this reading/writing in SerializableSourceText or here.  Do you prefer it here?

Reply:
I don't see a strong reason to move it right now. It's in the same location it was and doing essentially the same thing, plus related methods `DeserializeSerializableSourceText` and `DeserializeSourceText` appear together.

Reply:
sgtm.